rfd/0001-kubernetes-manager.md

```diff
@@ -0,0 +1,103 @@
+
+## Details
+
+### Implementation
+
+For the task, the [Kubebuilder](https://kubebuilder.io/) framework should be used. It has great code generation capabilities and gives a complete setup to develop & deploy the manager.
+
+### Teleport Auth server discovery and authentication
+
+To write something into Teleport, the manager needs to know what server to connect to and how. There're many options on how to discover Teleport instances in the cluster but for now, let's start with the simplest one - run manager as a sidecar container in the same pod as the Teleport auth server. Containers in the pod can share the volumes, so the manager container could have access to `/var/lib/teleport/proc` database, which contains administrator credentials. This is the same way how `tctl` utility works - it doesn't require any authentication when you run it on the same machine where Teleport is running.
+
+### Resource definitions
+
+Custom resource definitions of Teleport resource look like this:
+
+```yaml
+apiVersion: resources.goteleport.com/v2
+kind: Role
+metadata:
+  name: folks
+  namespace: teleport
+spec:
+  allow:
+    request:
+      roles: ["admin"]
+---
+apiVersion: resources.goteleport.com/v4
+kind: User
+metadata:
+  name: human-bean
+  namespace: teleport
+spec:
+  roles: ["folks"]
+```
+
+Manager process watches for custom resource events. Once a custom resource is created/updated, the manager's watcher acknowledges it and upserts a resource into Teleport.
+
+The name of the Kubernetes custom resource is the name of the resource in Teleport. Also, every Kubernetes custom resource must have a `spec` field that has the same schema as a `spec` field of a resource in Teleport.
+
+Deletion of resources is being handled using [resource finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/). Every acknowledged resource gets a finalizer named `"resources.goteleport.com/delete"` so when the deletion happens, Kubernetes doesn't actually delete it but sets a deletion timestamp. While the finalizer list is non-empty, the manager can still access the resource's fields. Deletion of the Kubernetes object leads to deletion of the associated resource in Teleport. Once a deletion in Teleport succeeds, the manager removes a finalizer from the Kubernetes object so it can be garbage-collected.
+
+### Identity definitions
+
+To run some plugin in the cluster, the credentials are required. Normally, one can get the credentials credentials using `tctl auth sign` command or using `tsh login`. To automate the credentials generation, the `Identity` resource is introduced. The identity definition looks like this:
+
+```yaml
+apiVersion: control.goteleport.com/v1
+kind: Identity
+metadata:
+  name: access-slack
+spec:
+  username: access-slack
+  secretName: access-slack-secret
```

Comment:
What I like about this RFD so far is that the CRDs are closely mapped to Teleport resources. That makes using the operator conceptually easy.

However, this resource breaks this pattern. "Identity" in this context is a new concept.

Both `tcl auth sign` and `tsh login` produce certificates under the hood, so I wonder if it might make more sense to call this a `Certificate` instead and replicate the flags of `tcl auth sign` in the `spec` section here.

Something like:

```yaml
apiVersion: resources.goteleport.com/v8
kind: Certificate
metadata:
  name: access-slack
spec:
  user: access-slack
  out:
    secretRef:
      name: access-slack-secret
```

Reply:
@pierrebeaucamp nice idea to unify the names!

However I cannot agree that `Certificate` is better. It's called an "identity file" in Teleport:

```
tctl auth sign --help
usage: tctl auth sign --out=OUT [<flags>]

Create an identity file(s) for a given user

Flags:
  -d, --debug              Enable verbose logging to stderr
  -c, --config             Path to a configuration file [/etc/teleport.yaml]. Can also be set via the TELEPORT_CONFIG_FILE environment variable.
      --auth-server        Attempts to connect to specific auth/proxy address(es) instead of local auth [127.0.0.1:3025]
  -i, --identity           Path to an identity file. Must be provided to make remote connections to auth. An identity file can be exported with 'tctl auth sign'
      --insecure           When specifying a proxy address in --auth-server, do not verify its TLS certificate. Danger: any data you send can be intercepted or modified by an
                           attacker.
      --user               Teleport user name
      --host               Teleport host name
  -o, --out                identity output
      --format             identity format: "file" (default), "openssh", "tls", "kubernetes", "db" or "mongodb"
      --ttl                TTL (time to live) for the generated certificate
      --compat             OpenSSH compatibility flag
      --proxy              Address of the teleport proxy. When --format is set to "kubernetes", this address will be set as cluster address in the generated kubeconfig file
      --overwrite          Whether to overwrite existing destination files. When not set, user will be prompted before overwriting any existing file.
      --leaf-cluster       Leaf cluster to generate identity file for when --format is set to "kubernetes"
      --kube-cluster-name  Kubernetes cluster to generate identity file for when --format is set to "kubernetes"
```

And in terms of PKI stuff, the identity file it's not just a certificate

Reply:
Let's keep calling it `Identity` then - thanks for pointing out that we already use that name in our CLI.

I don't mind keeping the `secretName` instead of `secretRef` either. I find the latter a bit easier to remember, but it's not a big deal really. Good to see that there is already some other CRDs calling it `secretName` as well. 👍